### PR TITLE
Added support for fetching gltf samples

### PR DESCRIPTION
**Pull request by xtanion**
This PR allows us to fetch and save glTF models directly from [Khronos gltf-samples](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0) repository.

Demo:
```py

from fury.data import fetch_viz_gltf, read_viz_gltf

fetch_viz_gltf(['BoxTextured', 'Fox'])

fname = read_viz_gltf('Fox')
print(fname)

```

### CODE REVIEW

**Comment by m-agour** [comment]
Hello @xtanion,
The fetcher works as it should, but I have two notes:

1. If the download stopped before downloading any of the three files, it will not download it again since the folder of the model exists (same if any of the files got deleted).
2. If tried to download a nonexisting model, the error `urllib.error.HTTPError: HTTP Error 404: Not Found` occurs while it should be an error showing that the model does not exist.

File: fury/data/fetcher.py [hunk 1]
```diff
@@ -502,3 +548,28 @@ def read_viz_dmri(fname):
...
+    Returns
+    --------
+    path : str
+        Complete path of models.
+    """
+    folder = pjoin(fury_home, 'glTF')
+    sample = pjoin(folder, fname)
+
+    if not os.path.exists(sample):
+        raise ValueError('Model doesnot exists.')
```
Comment (m-agour) [inline fury/data/fetcher.py:571]: Typo in `doesnot`

**Comment by xtanion** [comment]
Thanks for pointing that out @m-agour, I'll add a loop to check if all files exist and if not, It will download the files.

**Review by ganimtron-10 — COMMENTED** [review]
Hello @xtanion ,
The fetcher works fine but,
- As we can download multiple types of models (ie. glTF, glTF-binary, etc.). How do you differentiate and load the correct model type?
for eg. I downloaded the `glTF` and `glTF-binary` type of the `Fox` model but whenever I try to read the `Fox` model it only returns me the `.glTF` file(if exists) and doesn't give the `.glb` file.

**Comment by m-agour** [comment]
Hello @xtanion,
The fetcher works great. Now I think you should use this fetcher in the tutorial you made in the other glTF importer PR. Maybe rebase it on this PR and update the tutorial.

**Comment by xtanion** [comment]
> Hello @xtanion, The fetcher works great. Now I think you should use this fetcher in the tutorial you made in the other glTF importer PR. Maybe rebase it on this PR and update the tutorial.

Thanks @m-agour, I was thinking the same. I'll do that in a few days after it's ready to use `pygltflib`.

File: fury/data/fetcher.py [hunk 2]
```diff
@@ -267,6 +273,54 @@ def fetcher():
...
+    Parameter
+    -----------
+    name: str, list, optional
+        Name of the glTF model (for e.g. Box, BoxTextured, FlightHelmet, etc)
+        https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0
+        Default: None, Downloads essential glTF samples for tests.
+
+    mode: str, optional
+        Type of glTF format.
+        You can choose from different options (e.g. glTF, glTF-Embedded, etc)
```
Comment (skoudoro) [inline fury/data/fetcher.py:288]: instead of `etc` can you list the current possible mode

File: fury/data/fetcher.py [hunk 3]
```diff
@@ -267,6 +273,54 @@ def fetcher():
...
+        Default: glTF, `.bin` and texture files are stored separately.
+    """
+    if name is None:
+        name = ['BoxTextured', 'Duck', 'CesiumMilkTruck', 'CesiumMan']
+
+    if type(name) == list:
```
Comment (skoudoro) [inline fury/data/fetcher.py:294]: use `isinstance` instead

File: fury/data/fetcher.py [hunk 4]
```diff
@@ -267,6 +273,54 @@ def fetcher():
...
+        for element in name:
+            fetch_viz_gltf(element)
+    else:
+        url = GITHUB_API_URL + name + '/' + mode
```
Comment (skoudoro) [inline fury/data/fetcher.py:298]: use f-string please

File: fury/data/fetcher.py [hunk 5]
```diff
@@ -267,6 +273,54 @@ def fetcher():
...
+        try:
+            request = urlopen(url).read()
+            request = json.loads(request)
+        except HTTPError as e:
+            print("Model {} does not exist, Please check name and mode"
+                  .format(name))
+            return None
+
+        name = pjoin(* ['glTF', name, mode])
```
Comment (skoudoro) [inline fury/data/fetcher.py:307]: strange syntax, let's keep it simple

File: fury/data/fetcher.py [hunk 6]
```diff
@@ -267,6 +273,54 @@ def fetcher():
...
+            os.makedirs(folder)
+
+        for file in request:
+            download_url = file['download_url']
+            filename = download_url.split('/')[-1]
+            fullpath = os.path.join(folder, filename)
+
+            if not os.path.exists(fullpath):
+                print('Downloading file: {}'.format(filename))
+                urlretrieve(download_url, filename=fullpath)
```
Comment (skoudoro) [inline fury/data/fetcher.py:321]: as we said, let's use asyncio and ahttpio

File: fury/data/fetcher.py [hunk 7]
```diff
@@ -502,3 +556,54 @@ def read_viz_dmri(fname):
...
+
+    if not os.path.exists(sample):
+        raise ValueError('Model does not exists.')
+
+    for filename in os.listdir(sample):
+        if filename.endswith('.gltf') or filename.endswith('.glb'):
+            return pjoin(sample, filename)
+
+
+def list_gltf_samples():
```
Comment (skoudoro) [inline fury/data/fetcher.py:597]: `list_gltf_sample_models` instead

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

Thank you for this, Please, let us know when it is ready to be reviewed and merged

See below for some comments

File: fury/data/fetcher.py (hunk 6 above)
Comment (xtanion) [reply fury/data/fetcher.py:321]: YeAH, that needs to be implemented I am working on it, after that this PR will be ready to be merged.

**Comment by xtanion** [comment]
Hi @skoudoro, I made the changes, gltf fetcher is using `asyncio` and `aiohttp` now. The fetcher can be called by using
```py
asyncio.run(fetch_viz_gltf(['Fox', 'Box', 'Duck']))
```
PTAL, Thanks.

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

Thank you for the update. Please, be more rigorous in your docstring and in your way of coding in general. It will help a lot the reviewers to go to the essential.

See below for additional comments.

File: fury/data/fetcher.py [hunk 8]
```diff
@@ -267,6 +275,65 @@ def fetcher():
...
+async def _request(url, session):
+    async with session.get(url) as response:
+        return await response.json()
+
+
+async def _download(url, filename, session):
+    if not os.path.exists(filename):
+        print(f'Downloading: {filename}')
+        async with session.get(url) as response:
+            with open(filename, mode='wb') as f:
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/data/fetcher.py:287]: Hi @skoudoro , can I use `update_progressbar` instead? `copyfileobj_withprogress` is not an async function, and we need await to read the response.

[the remark this answers is not in the document]
Comment (skoudoro) [reply fury/data/fetcher.py:313]: ok

File: fury/data/fetcher.py [hunk 9]
```diff
@@ -267,6 +275,97 @@ def fetcher():
...
+    """
+    if not os.path.exists(filename):
+        print(f'Downloading: {filename}')
+        async with session.get(url) as response:
+            with open(filename, mode='wb') as f:
+                data = await response.read()
+                f.write(data)
+
+
+async def _get_viz_gltf(name, mode):
```
Comment (skoudoro) [inline fury/data/fetcher.py:318]: `_fetch_gltf` instead. 

do not forget the docstring...

File: fury/data/fetcher.py [hunk 10]
```diff
@@ -267,6 +275,97 @@ def fetcher():
...
+            d_urls = [file['download_url'] for file in request]
+            f_names = [pjoin(folder, url.split('/')[-1]) for url in d_urls]
+            zip_url = zip(d_urls, f_names)
+
+            await asyncio.gather(
+                *[_download(url, fname, session) for url, fname in zip_url]
+            )
+
+
+def fetch_viz_gltf(name=None, mode='glTF'):
```
Comment (skoudoro) [inline fury/data/fetcher.py:350]: `fetch_gltf` instead.  Why `viz`?

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

see below some additional comments

**Comment by skoudoro** [comment]
Also, do not forget to add the tests

**Comment by xtanion** [comment]
Thanks for the review @skoudoro, I'll add tests to the `test_gltf` in pr #600.

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

See below for some additional comments after testing the PR. 

Still some work to do before being merged. Let me know if you have any questions

File: fury/data/fetcher.py [hunk 11]
```diff
@@ -267,6 +275,97 @@ def fetcher():
...
+        https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0
+        Default: None, Downloads essential glTF samples for tests.
+
+    mode: str, optional
+        Type of glTF format.
+        You can choose from different options
+        (e.g. glTF, glTF-Embedded, glTF-Binary, glTF-Draco)
+        Default: glTF, `.bin` and texture files are stored separately.
+    """
+    asyncio.run(_get_viz_gltf(name, mode))
```
[the remark this answers is not in the document]
Comment (xtanion) [reply fury/data/fetcher.py:366]: Yeah, that's an issue. I need to add a check for name and model types. I'll do that.
I'll add tests in this PR, Should I create a  new file for tests, for eg: `tests_fetcher` or `test_gltf_fetcher`?

[the remark this answers is not in the document]
Comment (skoudoro) [reply fury/data/fetcher.py:366]: yes, new folder named `tests` in `fury.data`. Then a new file named test_fetcher.py

**Comment by xtanion** [comment]
Hi @skoudoro, I've added tests for fetching, reading, and validating gltf samples. `fetch_gltf` now returns a tuple with all fetched filenames. PTAL
Thanks

File: fury/data/fetcher.py [hunk 12]
```diff
@@ -12,7 +13,10 @@
 import tarfile
 import zipfile
 
-from urllib.request import urlopen
+from urllib.request import urlopen, urlretrieve
+from urllib.error import HTTPError
+import asyncio
+import aiohttp
```
Comment (skoudoro) [inline fury/data/fetcher.py:19]: pep8: `urlretrieve` and `httperror` not used

File: fury/data/tests/test_fetcher.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
+from fury.data import *
```
Comment (skoudoro) [inline fury/data/tests/test_fetcher.py:1]: We avoid as much as possible "import *".  Please, be more specific.

File: fury/data/tests/test_fetcher.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+import numpy as np
```
Comment (skoudoro) [inline fury/data/tests/test_fetcher.py:2]: pep8: not used

File: fury/data/tests/test_fetcher.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+import numpy.testing as npt
+import os
+from os.path import join as pjoin
+from aiohttp import InvalidURL
+import json
```
Comment (skoudoro) [inline fury/data/tests/test_fetcher.py:7]: pep8: import ordering.  check [here](https://peps.python.org/pep-0008/#imports) for more info

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

See below for my last comments. This PR should be ready to go after that. 
I also recommend you to activate pycodestyle and pydocstyle on your VS Code environment. Multiple error can be avoided after that

**Comment by skoudoro** [comment]
Also, Can you fix the conflict? thanks

**Comment by xtanion** [comment]
Hi @skoudoro, I made the changes and fixed the merge conflicts. PTAL
Thanks for your suggestion, I enabled `pycodestyle` & `pylint` in VS code, and they work really well.

File: fury/data/fetcher.py [hunk 16]
```diff
@@ -502,3 +569,54 @@ def read_viz_dmri(fname):
...
+
+    if not os.path.exists(sample):
+        raise ValueError('Model does not exists.')
+
+    for filename in os.listdir(sample):
+        if filename.endswith('.gltf') or filename.endswith('.glb'):
+            return pjoin(sample, filename)
+
+
+def list_gltf_sample_models():
```
[the remark this answers is not in the document]
Comment (skoudoro) [reply fury/data/fetcher.py:695]: hi @xtanion,

Where is the test concerning my comment above?

as a reminder

> The tests of this function is missing. Please, add `test_list_gltf_sample_models()`  in your `test_fetcher.py`

File: fury/data/tests/test_fetcher.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    npt.assert_raises(InvalidURL, fetch_gltf, ['duck'])
+    npt.assert_raises(InvalidURL, fetch_gltf, ['Duck'], 'GLTF')
+
+    items = os.listdir(boxtex)
+    npt.assert_array_equal(len(items), 3)
+
+    filenames, path = fetch_gltf('Box', 'glTF-Binary')
+    npt.assert_equal(len(filenames), 1)
+    print(filenames)
+    print(os.listdir(path))
```
Comment (skoudoro) [inline fury/data/tests/test_fetcher.py:38]: Can you remove those print

**Review by skoudoro — COMMENTED** [review]
Hi @xtanion,

See the comment below.

File: fury/data/tests/test_fetcher.py (hunk 17 above)
Comment (xtanion) [reply fury/data/tests/test_fetcher.py]: Sorry I missed that comment. Added tests and removed those prints.

**Pull request merged**